File: Platform.Collections.Methods/Trees/SizeBalancedTreeMethods.cs
using System;

#pragma warning disable CS1591 // Missing XML comment for publicly visible type or member

namespace Platform.Collections.Methods.Trees
{
    public abstract class SizeBalancedTreeMethods<TElement> : SizedBinaryTreeMethodsBase<TElement>
    {
        protected override void AttachCore(ref TElement root, TElement node)
        {
            if (EqualToZero(root))
            {
                root = node;
                IncrementSize(root);
            }
            else
            {
                IncrementSize(root);
                if (FirstIsToTheLeftOfSecond(node, root))
                {
                    AttachCore(ref GetLeftReference(root), node);
                    LeftMaintain(ref root);
                }
                else
                {
                    AttachCore(ref GetRightReference(root), node);
                    RightMaintain(ref root);
                }
            }
        }

        protected override void DetachCore(ref TElement root, TElement nodeToDetach)
        {
            ref var currentNode = ref root;
            ref var parent = ref root;
            var replacementNode = Zero;
            while (!AreEqual(currentNode, nodeToDetach))
            {
                DecrementSize(currentNode);
                if (FirstIsToTheLeftOfSecond(nodeToDetach, currentNode))
                {
                    parent = ref currentNode;
                    currentNode = ref GetLeftReference(currentNode);
                }
                else if (FirstIsToTheRightOfSecond(nodeToDetach, currentNode))
                {
                    parent = ref currentNode;
                    currentNode = ref GetRightReference(currentNode);
                }
                else
                {
                    throw new InvalidOperationException("Duplicate link found in the tree.");
                }
            }
            var nodeToDetachLeft = GetLeft(nodeToDetach);
            var node = GetRight(nodeToDetach);
            if (!EqualToZero(nodeToDetachLeft) && !EqualToZero(node))
            {
                var leftestNode = GetLeftest(node);
                DetachCore(ref GetRightReference(nodeToDetach), leftestNode);
                SetLeft(leftestNode, nodeToDetachLeft);
                node = GetRight(nodeToDetach);
                if (!EqualToZero(node))
                {
                    SetRight(leftestNode, node);
                    SetSize(leftestNode, Increment(Add(GetSize(nodeToDetachLeft), GetSize(node))));
                }
                else
                {
                    SetSize(leftestNode, Increment(GetSize(nodeToDetachLeft)));
                }
                replacementNode = leftestNode;
            }
            else if (!EqualToZero(nodeToDetachLeft))
            {
                replacementNode = nodeToDetachLeft;
            }
            else if (!EqualToZero(node))
            {
                replacementNode = node;
            }
            if (AreEqual(root, nodeToDetach))
            {
                root = replacementNode;
            }
            else if (AreEqual(GetLeft(parent), nodeToDetach))
            {
                SetLeft(parent, replacementNode);
            }
            else if (AreEqual(GetRight(parent), nodeToDetach))
            {
                SetRight(parent, replacementNode);
            }
            ClearNode(nodeToDetach);
        }

        private void LeftMaintain(ref TElement root)
        {
            if (!EqualToZero(root))
            {
                var rootLeftNode = GetLeft(root);
                if (!EqualToZero(rootLeftNode))
                {
                    var rootRightNode = GetRight(root);
                    var rootRightNodeSize = GetSize(rootRightNode);
                    var rootLeftNodeLeftNode = GetLeft(rootLeftNode);
                    if (!EqualToZero(rootLeftNodeLeftNode) &&
                        (EqualToZero(rootRightNode) || GreaterThan(GetSize(rootLeftNodeLeftNode), rootRightNodeSize)))
                    {
                        RightRotate(ref root);
                    }
                    else
                    {
                        var rootLeftNodeRightNode = GetRight(rootLeftNode);
                        if (!EqualToZero(rootLeftNodeRightNode) &&
                            (EqualToZero(rootRightNode) || GreaterThan(GetSize(rootLeftNodeRightNode), rootRightNodeSize)))
                        {
                            LeftRotate(ref GetLeftReference(root));
                            RightRotate(ref root);
                        }
                        else
                        {
                            return;
                        }
                    }
                    LeftMaintain(ref GetLeftReference(root));
                    RightMaintain(ref GetRightReference(root));
                    LeftMaintain(ref root);
                    RightMaintain(ref root);
                }
            }
        }

        private void RightMaintain(ref TElement root)
        {
            if (!EqualToZero(root))
            {
                var rootRightNode = GetRight(root);
                if (!EqualToZero(rootRightNode))
                {
                    var rootLeftNode = GetLeft(root);
                    var rootLeftNodeSize = GetSize(rootLeftNode);
                    var rootRightNodeRightNode = GetRight(rootRightNode);
                    if (!EqualToZero(rootRightNodeRightNode) &&
                        (EqualToZero(rootLeftNode) || GreaterThan(GetSize(rootRightNodeRightNode), rootLeftNodeSize)))
                    {
                        LeftRotate(ref root);
                    }
                    else
                    {
                        var rootRightNodeLeftNode = GetLeft(rootRightNode);
                        if (!EqualToZero(rootRightNodeLeftNode) &&
                            (EqualToZero(rootLeftNode) || GreaterThan(GetSize(rootRightNodeLeftNode), rootLeftNodeSize)))
                        {
                            RightRotate(ref GetRightReference(root));
                            LeftRotate(ref root);
                        }
                        else
                        {
                            return;
                        }
                    }
                    LeftMaintain(ref GetLeftReference(root));
                    RightMaintain(ref GetRightReference(root));
                    LeftMaintain(ref root);
                    RightMaintain(ref root);
                }
            }
        }
    }
}

File: cpp/Platform.Collections.Methods/SizeBalancedTreeMethods.cpp
namespace Platform::Collections::Methods::Trees
{
    template <typename TElement> class SizeBalancedTreeMethods : public SizedBinaryTreeMethodsBase<TElement>
    {
    public:
        void AttachCore(TElement* root, TElement node) override
        {
            if (*root == 0)
            {
                *root = node;
                this->IncrementSize(*root);
            }
            else
            {
                this->IncrementSize(*root);
                if (this->FirstIsToTheLeftOfSecond(node, *root))
                {
                    this->AttachCore(this->GetLeftReference(*root), node);
                    this->LeftMaintain(root);
                }
                else
                {
                    this->AttachCore(this->GetRightReference(*root), node);
                    this->RightMaintain(root);
                }
            }
        }

        void DetachCore(TElement* root, TElement nodeToDetach) override
        {
            auto* currentNode = root;
            auto* parent = root;
            auto replacementNode = 0;
            while (*currentNode != nodeToDetach)
            {
                this->DecrementSize(*currentNode);
                if (this->FirstIsToTheLeftOfSecond(nodeToDetach, *currentNode))
                {
                    parent = currentNode;
                    currentNode = this->GetLeftReference(*currentNode);
                }
                else if (this->FirstIsToTheRightOfSecond(nodeToDetach, *currentNode))
                {
                    parent = currentNode;
                    currentNode = this->GetRightReference(*currentNode);
                }
                else
                {
                    throw std::runtime_error("Duplicate link found in the tree.");
                }
            }
            auto nodeToDetachLeft = this->GetLeft(nodeToDetach);
            auto node = this->GetRight(nodeToDetach);
            if (nodeToDetachLeft != 0 && node != 0)
            {
                auto leftestNode = this->GetLeftest(node);
                this->DetachCore(this->GetRightReference(nodeToDetach), leftestNode);
                this->SetLeft(leftestNode, nodeToDetachLeft);
                node = this->GetRight(nodeToDetach);
                if (node != 0)
                {
                    this->SetRight(leftestNode, node);
                    this->SetSize(leftestNode, (this->GetSize(nodeToDetachLeft) + this->GetSize(node)) + 1);
                }
                else
                {
                    this->SetSize(leftestNode, this->GetSize(nodeToDetachLeft) + 1);
                }
                replacementNode = leftestNode;
            }
            else if (nodeToDetachLeft != 0)
            {
                replacementNode = nodeToDetachLeft;
            }
            else if (node != 0)
            {
                replacementNode = node;
            }
            if (*root == nodeToDetach)
            {
                *root = replacementNode;
            }
            else if (this->GetLeft(*parent) == nodeToDetach)
            {
                this->SetLeft(*parent, replacementNode);
            }
            else if (this->GetRight(*parent) == nodeToDetach)
            {
                this->SetRight(*parent, replacementNode);
            }
            this->ClearNode(nodeToDetach);
        }

        void LeftMaintain(TElement* root)
        {
            if (*root != 0)
            {
                auto rootLeftNode = this->GetLeft(*root);
                if (rootLeftNode != 0)
                {
                    auto rootRightNode = this->GetRight(*root);
                    auto rootRightNodeSize = this->GetSize(rootRightNode);
                    auto rootLeftNodeLeftNode = this->GetLeft(rootLeftNode);
                    if (rootLeftNodeLeftNode != 0 &&
                        (rootRightNode == 0 || this->GetSize(rootLeftNodeLeftNode) > rootRightNodeSize))
                    {
                        this->RightRotate(root);
                    }
                    else
                    {
                        auto rootLeftNodeRightNode = this->GetRight(rootLeftNode);
                        if (rootLeftNodeRightNode != 0 &&
                            (rootRightNode == 0 || this->GetSize(rootLeftNodeRightNode) > rootRightNodeSize))
                        {
                            this->LeftRotate(this->GetLeftReference(*root));
                            this->RightRotate(root);
                        }
                        else
                        {
                            return;
                        }
                    }
                    this->LeftMaintain(this->GetLeftReference(*root));
                    this->RightMaintain(this->GetRightReference(*root));
                    this->LeftMaintain(root);
                    this->RightMaintain(root);
                }
            }
        }

        void RightMaintain(TElement* root)
        {
            if (*root != 0)
            {
                auto rootRightNode = this->GetRight(*root);
                if (rootRightNode != 0)
                {
                    auto rootLeftNode = this->GetLeft(*root);
                    auto rootLeftNodeSize = this->GetSize(rootLeftNode);
                    auto rootRightNodeRightNode = this->GetRight(rootRightNode);
                    if (rootRightNodeRightNode != 0 &&
                        (rootLeftNode == 0 || this->GetSize(rootRightNodeRightNode) > rootLeftNodeSize))
                    {
                        this->LeftRotate(root);
                    }
                    else
                    {
                        auto rootRightNodeLeftNode = this->GetLeft(rootRightNode);
                        if (rootRightNodeLeftNode != 0 &&
                            (rootLeftNode == 0 || this->GetSize(rootRightNodeLeftNode) > rootLeftNodeSize))
                        {
                            this->RightRotate(this->GetRightReference(*root));
                            this->LeftRotate(root);
                        }
                        else
                        {
                            return;
                        }
                    }
                    this->LeftMaintain(this->GetLeftReference(*root));
                    this->RightMaintain(this->GetRightReference(*root));
                    this->LeftMaintain(root);
                    this->RightMaintain(root);
                }
            }
        }
    };
}